allennlp/modules/token_embedders/embedding.py

```diff
@@ -99,6 +102,7 @@ def __init__(
         sparse: bool = False,
         vocab_namespace: str = None,
         pretrained_file: str = None,
+        vocabulary: Vocabulary = None,
```

Comment:
If you're taking this parameter in the constructor, I think you can just remove `from_vocab_or_files` entirely, and remove the `constructor=` parameter from the registry, and the duplicate registration at the bottom of the class.  I'm in favor of this, it simplifies things quite a bit.

Reply:
Yes, I thought a bit about this. However:

1. programmatic usage should still support `Embedding(num_embeddings, embedding_dim)`. If I make num_embeddings optional (as it can now be inferred if a vocab is also passed), I will have to break Embedding in a really insidious way, as it is the first argument. 

2. There is some interplay with `vocab_namespace` being set to `None` if `num_embeddings` is not passed and how the vocabulary is extended which I don't really understand.

Reply:
1. You'll only be breaking it for people who called it directly, which is likely a minority of users.  We can have a big warning about this in our release notes.  It does seem like a pretty big reduction in complexity whose only cost is a breaking change.

2. I think the main complexity is that the default in `from_vocab_or_files` is `"tokens"`, while it's `None` here.  That code block was trying to override the default value.  It'd be pretty easy to just do `num_embeddings = num_embeddings or vocab.get_vocab_size(vocab_namespace or "tokens")` in the constructor (maybe not so one-line, but that's the idea), and there's no issue.

You can pick whether you think it's worth the breaking change or not.  I think it probably is, but I'm ok if you think otherwise.

Reply:
Please, only one request. Do not forget the people who use the lib directly, not using the config. It is a great use case for production apps who need to process in-memory data. The Flair library do an awesome work on this area. But I really love the AllenNLP maturity. ❤️

Reply:
A major effort in our upcoming 1.0 release is to make everything easier to use for people who don't want to use our config files. So, no, we haven't forgotten you :). In fact, this recommended breakage is specifically to make it easier in the long run to use this module without config files, it just requires a one-time breaking change to make it better.